src/Core/Content/Product/SalesChannel/Detail/ProductDetailRoute.php

```diff
@@ -114,6 +124,10 @@ public function load(string $productId, Request $request, SalesChannelContext $c
 
         $criteria->setIds([$productId]);
 
+        $this->eventDispatcher->dispatch(
```

Comment:
It's not intended that the Store-API Routes have a Criteria Event. The PHP caller should have one, as it passes it into it.
Also this makes the Caching Workflow much complicated

Reply:
recently I also wanted to add an event in a route, but as Soner already said, this should not be done. 

Another reason is, that the routes are also called via API and with that change you will force an API caller to load all the data, you add with this criteria event, also maybe the caller wanted his payload as small as possible. 

that is the reason why the the event should happen before the route is called

Reply:
Just humor me then, how should i implement a custom closeout handling for a customer? And some more questions i really like answer :D

It's not intended that the Store-API Routes have a Criteria Event: where is this documented?
The PHP caller should have one: Then why is the criteria changed inside the store-api-routes?
you will force an API caller to load all the data: Where does an event force the caller to load all the data?

Reply:
the API caller has no chance to switch off the data, which is added with this event. unless he/she also subscribes to this event :wink: that is what I meant.

the ProductDetailRoute is a bad example :see_no_evil: here happens to much for a route class. this needs some refactoring for sure. am I right @OliverSkroblin ?

Reply:
"the API caller has no chance to switch off the data": The api caller also does not have the possibility to change the system-config value 🤔 Can you elaborate on your thoughts in that regard?

Also: An API-Caller should never be able to change the business logic, in my case what happens in my subscibers to these events, at all. Imagine you could change the internal workings of the amazon store api and "switch off data". The business logic should never be exposed.

Reply:
"the ProductDetailRoute is a bad example" if the routes needs heavy refactoring, why not add an event in the meantime so some can change the logic for their current business needs? When the big refactoring happens you surely break other stuff so removing/moving the event is not really a deal breaker ;)

Reply:
@shyim @mitelg could you answer my other questions, too? :D

Reply:
what I meant is, that a plugin could add tens of associations, and the API caller wonders about all the data, he/she does not need. I didn't talk about business logic :slightly_smiling_face:

Reply:
@mitelg but the api callers for my shopware installation are in my direct responsibility. We are not talking about a public api like amazon. If my business logic changes some associations and some third party systems of me breaks because of that, its not the problem of shopware as a software provider - its my problem to deal with my third party integrations.

Reply:
Hey @jochenmanz,

so I have discussed the whole thing internally. 

Events, to adjust the criteria, should always take place in the corresponding page loaders, because the developers can make the adjustments more selectively and not drag down the performance of the whole system.

If you want to add business logic for the products. Let's say as an example an own AvailableFilter, you should better decorate the SalesChannelProductDefinition. This way you have directly implemented all the places.

Reply:
@OliverSkroblin when will all the criteria changes be moved to the pageloaders then? Any chance that can be implemented in the next minor Version? 

Because currently our new logic is page independent, but i do have some tickets in the future where new business logic needs to be added depending on the current customer, customergroup and route ;)

Reply:
@shyim i've removed the event from this PR